cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -318,6 +319,32 @@ private void getScheduleStatus(HttpResponder responder, String appId, String nam
     }
   }
 
+  /**
+   * Stops the particular run of the Workflow or MapReduce program.
+   */
+  @POST
+  @Path("/apps/{app-id}/{type}/{id}/runs/{run-id}/stop")
+  public void performRunLevelStop(HttpRequest request, HttpResponder responder,
+                                  @PathParam("namespace-id") String namespaceId,
+                                  @PathParam("app-id") String appId,
+                                  @PathParam("type") String type,
+                                  @PathParam("id") String id,
+                                  @PathParam("run-id") String runId) {
+    try {
+      ProgramType programType = ProgramType.valueOfCategoryName(type);
+      if (!isRunLevelActionAllowed(programType)) {
+        responder.sendString(HttpResponseStatus.FORBIDDEN,
```

Comment:
On second thoughts, this may be unrelated to this PR, so I think it is ok to report a jira and handle in a separate PR. Even more so, because I think these exceptions do not have a standard HttpResponse currently. They either result in ``FORBIDDEN`` or ``CONFLICT``. Definitely a separate issue from this one.